vmdb/app/models/ems_refresh/parsers/kubernetes.rb

```diff
@@ -230,6 +233,16 @@ def parse_container_port_config(port_config, pod_id, container_name)
       }
     end
 
+    def parse_service_port_config(port_config, service_id)
+      {
+        :ems_ref     => "#{service_id}_#{port_config["port"]}_#{port_config["targetPort"]}_#{port_config["protocol"]}",
```

Comment:
You need some sort of unique identifier otherwise the save inventory code
will thrash on every refresh (every record will be an add and a delete).
You can either use ems_ref or have a multi-value key in the save inventory.

ems_ref is technically supposed to be the actual value for an entity from
the EMS, so you can communicate that value back to the native API if
needed.  However we've built keys like this before if it didn't make sense,
or a real value isn't available.

The choice for having ems_ref versus a multi key is also driven by multiple
providers.   In this case, what is the equivalent for Amazon containers,
and does that equivalent have a unique identifier.  If so, then you have to
have ems_ref; if not then either way is fine.

/cc @blomquisg
 On Apr 7, 2015 7:13 AM, "abonas" <notifications@github.com> wrote:

> In vmdb/app/models/ems_refresh/parsers/kubernetes.rb
> <https://github.com/ManageIQ/manageiq/pull/2523#discussion_r27868681>:
>
> > @@ -230,6 +233,16 @@ def parse_container_port_config(port_config, pod_id, container_name)
> >        }
> >      end
> >
> > +    def parse_service_port_config(port_config, service_id)
> > +      {
> > +        :ems_ref     => "#{service_id}_#{port_config["port"]}_#{port_config["targetPort"]}_#{port_config["protocol"]}",
>
>  Maybe in this case ems_ref can be omitted altogether:
>
> service ports are not an entity on their own (they are just a one-to-many items of services)
> using all the attributes as ems_ref basically means that there's no field that can be updated without generating a new entity
>
> If save_inventory_multi supports it I'd suggest to remove the ems_ref in
> this specific case. (/cc @Fryguy <https://github.com/Fryguy> )
>
> this is similar to what was done in containers/containers def/container
> port configs - they all don't have a "real" ems_ref. iirc, it was advised
> by @Fryguy <https://github.com/Fryguy> to have those ems_ref fields
> generated (@Fryguy <https://github.com/Fryguy> pls correct me if I'm
> wrong).
> I'm OK to drop them if they are not needed. And if we're gonna drop them
> -imo it has to be done for all the "nested" entities.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ManageIQ/manageiq/pull/2523/files#r27868681>.
>